### ISSUE TITLE
Analyze perf impact of moving setStartupMetrics earlier

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/fenix#19395]
In https://github.com/mozilla-mobile/fenix/issues/19147, in order to work around an issue with start up metrics appearing null, we moved their assignments earlier. Let's analyze the performance impact.

**Comment by mcomella** [comment]
tldr; I think this creates a regression of ~114ms or -8.07% on the Moto G5. It appears in our time to visual completeness on homescreen test and I'd expect a similar impact on COLD VIEW (app link) but that's not what we measured (but that system has issues).

**This is a very large regression.** In my opinion, there's no reason collecting metrics should non-trivially slow down the user to seeing the first frame or loading a page and I think we can do better. For example, might we:
- move the collection/submission of the Glean ping later so we can set the start up metrics later?
- capture these off the main thread? It creates a possible race and introduces complexity (can all of these be captured off the main thread?) so it doesn't sound ideal

### Details
I ran this in FNPRMS (i.e. time from process start to homescreen visual completeness) 25 iterations on my Moto G5 (mean?):
- before: 1.4109600000000002s
- after: 1.5248399999999998s

That's a **regression of 114ms, or -8.07% difference.**

mozperftest VIEW (i.e. time from process start to gecko navigationStart) is untrustworthy right now (I introduced an intentional 100ms thread.sleep and saw an improvement) but I ran it too, 50 iterations (median):
- before: 2215.5ms
- after: 2141.0ms

It's possible this 74.5ms improvement is real – perhaps we're reducing the load on the disk during page load (`setStartupMetrics` ran on the visual completeness queue before) – but it seems more likely to be a perftest noise.

I took **profiles** to better understand the impact:
- before: https://share.firefox.dev/3b10D7F
- after: https://share.firefox.dev/3nY0EyM

This profiler (`Debug` API) has noticeable overhead so the exact times don't mean much and we're not sure if the relative times mean much but the duration of `FenixApplication.onCreate` regressed 25.32%. This could also be a fluke: I notice there's an extra long wait at the end of `onCreate` in the after profile but perhaps that is caused by the change.

The after profile is hard to interpret because `setStartupMetrics` gets inlined but it creates a noticeable change in the profile – all the small segments that occur over the course of 200ms seem to be introduced:
![image](https://user-images.githubusercontent.com/759372/117183316-420b5e00-ad8c-11eb-971e-8307b4ba5aa1.png)

Without spending more time, I'm not sure how to further interpret this.

**Comment by mcomella** [comment]
I've done the analysis and will put this into waiting while we wait to see how to address the regression or if the product team is okay accepting it.

**Comment by jonalmeida** [comment]
I left a suggestion in https://github.com/mozilla-mobile/fenix/pull/19397#issuecomment-833965318 that you might be interested in as well.

**Comment by mcomella** [comment]
The fix landed in https://github.com/mozilla-mobile/fenix/pull/19397#issuecomment-833965318: let's verify it's improved roughly back to the way it was. Let's use FNPRMS MAIN like I did above. We should measure 3 things: the original performance, after the regression, and after the fix.

Note: jonalmeida ran it through the profiler and didn't see issues https://github.com/mozilla-mobile/fenix/pull/19397#pullrequestreview-657417016 but we should measure it with timestamps too.

**Comment by mcomella** [comment]
I re-ran the numbers after the async fix. For time to first frame COLD VIEW, I see a 53ms regression (-3.55%) from before any change landed to after this fix landed. Here are the medians in ms:
- before: 1494.5
- after regression: 1607.0
- after fix: 1547.5

And here's the replicate + analysis data:
- [before-analysis.txt](https://github.com/mozilla-mobile/fenix/files/6587921/before-analysis.txt)
- [regression-analysis.txt](https://github.com/mozilla-mobile/fenix/files/6587924/regression-analysis.txt)
- [fix-analysis.txt](https://github.com/mozilla-mobile/fenix/files/6587927/fix-analysis.txt)

**Comment by mcomella** [comment]
> I re-ran the numbers after the async fix. For time to first frame COLD VIEW, I see a 53ms regression (-3.55%) from before any change landed to after this fix landed.

I filed https://github.com/mozilla-mobile/fenix/issues/20346 since afaict, no fix has been made and its probably because this was forgotten about because there is no issue representing it.

I'll close this bug (since we analyzed the impact) and we'll do the remaining work in the new bug.
